### PR TITLE
Abstract syscall/js package for internal usage

### PR DESCRIPTION
**Pull request by DeeStarks**

### CODE REVIEW

File: middleware.go [hunk 1]
```diff
@@ -43,117 +38,47 @@ func main() {
...
+	)
 
 	// proceed to next middleware/handler request is not a layer8 request
 	if headers.String() == "<undefined>" || headers.Get("x-tunnel").String() == "<undefined>" {
 		next.Invoke()
 		return nil
 	}
+	
+	initECDH := func() interface{} {
+		secret, pub, mpjwt, err := internals.InitializeECDH(goHeaders)
```
Comment (stravid87) [inline middleware.go:60]: What prevents you from calling internals.InitializeECDH(goHeaders) directly here? Instead of putting inside a function?

File: internals/process_data.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"globe-and-citizen/layer8/middleware/js"
+
+	utils "github.com/globe-and-citizen/layer8-utils"
+)
+
+// ProcessData decrypts the request body and processes it. It then sets the method and
```
Comment (stravid87) [inline internals/process_data.go:15]: Nice comments!

File: middleware.go [hunk 3]
```diff
@@ -194,197 +117,91 @@ func WASMMiddleware_v2(this js.Value, args []js.Value) interface{} {
...
-						)
-						formData.Call("append", k, file)
-					case "String":
-						formData.Call("append", k, val["value"].(string))
-					case "Number":
-						formData.Call("append", k, val["value"].(float64))
-					case "Boolean":
-						formData.Call("append", k, val["value"].(bool))
+		var (
+			formdata   = js.Global().Get("FormData").New()
```
Comment (stravid87) [inline middleware.go:121]: You create goFormData. Line 127 - 152 then complete it prior to it being passed in at line 153. If internals.processData is inspected however, this function only uses processData IF `case "application/layer8.buffer+json":`. Is there a way to refactor this to make it cleaner?

Comment (stravid87) [reply middleware.go:121]: The majority of all calls will likely be case "application/Json"

File: internals/prepare_data.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	b, err = symmKey.SymmetricEncrypt(b)
+	if err != nil {
+		println("error encrypting response:", err.Error())
+		return &utils.Response{
+			Status:     500,
+			StatusText: "Could not encrypt response",
+		}
+	}
+
+	resHeaders := make(map[string]interface{})
```
Comment (stravid87) [inline internals/prepare_data.go:82]: What is the purpose of resHeaders here?

File: js/marshal.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+// Unmarshal unmarshals a `syscall/js.Value` into the internal `Value` type.
+//
+// Note that only primitive types are supported and will be unmarshaled
+//   - Number
+//   - Boolean
+//   - String
+//   - Object (recursively converted to map[string]interface{})
+//   - Array (recursively converted to []interface{})
+func Unmarshal(v js.Value) *Value {
```
Comment (stravid87) [inline js/marshal.go:143]: This will need to be an interface... otherwise, we're going to get the same build error during testing :(

File: js/marshal.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+func (v Value) Get(key string) interface{} {
+	val, ok := v.value.(map[string]interface{})[key]
+	if !ok {
+		return nil
+	}
+	return val.(Value).value
+}
+
+func (v Value) Set(key string, value interface{}) {
```
Comment (stravid87) [inline js/marshal.go:79]: What happens if?

```
myInt : = 99
newVal := js.NewValue(myInt) // that is I pass in an int or other non-object
newVal.Set(<any-key-string>, <any-other-value>)
```

**Review by stravid87 — CHANGES_REQUESTED** [review]
Nice work. Took me 3 hours but here are 7 comments... 
Unfortunately still some work todo but we're getting closer and each refactor makes the next  easier to understand and reason about.

File: middleware.go (hunk 1 above)
Comment (DeeStarks) [reply middleware.go:60]: It is for reusability. There are many blocks in the code that call `InitializeECDH`, so instead of writing the code inside this function for each of those blocks (which is repetition), I put it in this function and call it when I need to `InitializeECDH` without having to rewrite the code

File: middleware.go (hunk 3 above)
Comment (DeeStarks) [reply middleware.go:121]: > If internals.processData is inspected however, this function only uses processData IF `case "application/layer8.buffer+json"`:

This is not the case, every request goes through ProcessData and before the processing, it checks for the content type. When it is `application/layer8.buffer+json`, that's when it uses the form data interface, if not (that's the default case in the switch statement), it no longer needs to do any further processes as what's necessary has been handled in the starting section of the function. So, both `multipart/formdata` and `application/json` are handled properly. 
I've tested both and confirmed they both work

File: internals/prepare_data.go (hunk 4 above)
Comment (DeeStarks) [reply internals/prepare_data.go:82]: Good catch! It's not used. I'll remove it

File: js/marshal.go (hunk 5 above)
Comment (DeeStarks) [reply js/marshal.go:143]: Yes, I didn't realize that initially. However with an interface, we'll still come back to it, as we'll need to assert

File: js/marshal.go (hunk 6 above)
Comment (DeeStarks) [reply js/marshal.go:79]: It'll fail. But that's a good one, instead of panicking, we could fail silently by returning the error to the caller and allowing them to handle it

Comment (DeeStarks) [reply js/marshal.go]: Actually, this is not the case. It type asserts and properly sets the value as an `interface{}` type

File: internals/init_ecdh_test.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		wantShared    string
+		wantPub       string
+		wantMpJWT     string
+		wantErr       bool
+		wantErrString string
+	}{
+		{
+			name: "init_with_valid_headers",
+			args: args{
+				headers: js.ValueOf(map[string]interface{}{
```
Comment (stravid87) [inline internals/init_ecdh_test.go:64]: To do the testing systematically, my understanding is that the argument to InitializeECDH() becomes the source of variance. It is a map[string]interface{} which implies the following test cases:
```
EMPTY: 
headers: js.ValueOf(map[string]interface{}{}

RANDOM ADDED HEADERS / TOO MANY HEADERS:
headers: js.ValueOf(map[string]interface{}{
    "addedheader": "artibtrarystring"
    "x-ecdh-init": b64ClientPub,
    "x-client-uuid": uuid.New().String(),
     "mp-jwt": mpjwt,
}

DUPLICATE HEADERS:
headers: js.ValueOf(map[string]interface{}{
    "x-ecdh-init": b64ClientPub,
    "x-ecdh-init": b64ClientPub,
    "x-client-uuid": uuid.New().String(),
     "mp-jwt": mpjwt,
}


AND FOR EACH OF THE STRINGS, STRING ABUSES:
headers: js.ValueOf(map[string]interface{}{
     "x-ecdh-init":   b64ClientPub | invalid characters | <verylongstring> | <multiline\nstring> | ""
     "x-client-uuid": uuid.New().String() | invalid characters | <verylongstring> | <multiline\nstring> | ""
     "mp-jwt":        mpjwt | invalid characters | <verylongstring> |  <multiline\nstring> | ""
}
```

**Review by stravid87 — CHANGES_REQUESTED** [review]
Nice work on the PR  -- there is lots of good work added here. 
The feedback here is based on my understanding of best practices as it comes from the testing Go Book I was reading. Mainly, focus on the function under test (i.e., single responsibility) and systematically test standard edge cases -- I'll post a picture for reference in the chat section.

File: internals/init_ecdh_test.go (hunk 7 above)
Comment (stravid87) [reply internals/init_ecdh_test.go:64]: image

File: internals/init_ecdh_test.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	"testing"
+
+	utilities "github.com/globe-and-citizen/layer8-utils"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestInitializeECDH(t *testing.T) {
+	serverPri, serverPub, err := utilities.GenerateKeyPair(utilities.ECDH)
+	assert.Nil(t, err)
```
[the remark this answers is not in the document]
Comment (DeeStarks) [reply internals/init_ecdh_test.go:16]: Yes, these assertions are to ensure that requirements are met for what we're testing for. Indeed they should be tested in the utilities but the code coverage here compared to what the test would look like in the utilities package would be different
This is simply just checking that what we need for the current package we're testing for is provided

File: internals/init_ecdh_test.go (hunk 7 above)
Comment (DeeStarks) [reply internals/init_ecdh_test.go:64]: Good point! I'll add these cases

File: internals/init_ecdh_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (stravid87) [reply internals/init_ecdh_test.go:16]: K sounds good.

**Review by stravid87 — APPROVED** [review]
